arch/arm/core/elf.c

```diff
@@ -11,42 +12,261 @@
+#define SHIFT_JUMPS_SIGN 25
+#define SHIFT_MOV_RD 4
+#define SHIFT_MOV_RN 4
+#define SHIFT_MOVS_SIGN 15
+#define SHIFT_THM_JUMPS_SIGN 24
+#define SHIFT_THM_BW_IMM10 12
+#define SHIFT_THM_BL_J2 22
+#define SHIFT_THM_BL_J1 23
+#define SHIFT_THM_MOVS_SIGN 15
+#define SHIFT_THM_MOV_I 1
+#define SHIFT_THM_MOV_IMM3 4
+#define SHIFT_THM_MOV_IMM4 12
 
-static int32_t arm_bl_blx_decode_addend(uintptr_t opaddr)
+static inline int prel31_decode(elf_word reloc_type, uint32_t loc,
+				uint32_t sym_base_addr, const char *sym_name, int32_t *offset)
 {
-	uint16_t upper = *((uint16_t *)opaddr);
-	uint16_t lower = *(((uint16_t *)opaddr) + 1);
+	int ret;
 
-	int32_t addend = upper & ARM_BL_BLX_UPPER_S_BIT ? UINT32_MAX : 0;
+	*offset = sign_extend(*(int32_t *)loc, SHIFT_PREL31_SIGN);
+	*offset += sym_base_addr - loc;
+	if (*offset >= PREL31_UPPER_BOUNDARY || *offset < PREL31_LOWER_BOUNDARY) {
+		LOG_ERR("sym '%s': relocation out of range (%#x -> %#x)\n",
+			sym_name, loc, sym_base_addr);
+		ret = -ENOEXEC;
+	} else {
+		ret = 0;
+	}
+
+	return ret;
+}
+
+static inline void prel31_reloc(uint32_t loc, int32_t *offset)
+{
+	*(uint32_t *)loc &= BIT(31);
+	*(uint32_t *)loc |= *offset & GENMASK(30, 0);
+}
+
+static int prel31_handler(elf_word reloc_type, uint32_t loc,
+				uint32_t sym_base_addr, const char *sym_name, int32_t *offset)
+{
+	int ret;
+
+	ret = prel31_decode(reloc_type, loc, sym_base_addr, sym_name, offset);
+	if (!ret) {
+		prel31_reloc(loc, offset);
+	}
+
+	return ret;
+}
+
+static inline int jumps_decode(elf_word reloc_type, uint32_t loc,
+				uint32_t sym_base_addr, const char *sym_name, int32_t *offset)
+{
+	int ret;
+
+	*offset = MEM2ARMOPCODE(*(uint32_t *)loc);
+	*offset = (*offset & MASK_BRANCH_OFFSET) << SHIFT_BRANCH_OFFSET;
+	*offset = sign_extend(*offset, SHIFT_JUMPS_SIGN);
+	*offset += sym_base_addr - loc;
+	if (*offset <= JUMP_UPPER_BOUNDARY || *offset >= JUMP_LOWER_BOUNDARY) {
+		LOG_ERR("sym '%s': relocation out of range (%#x -> %#x)\n",
+			sym_name, loc, sym_base_addr);
+		ret = -ENOEXEC;
+	} else {
+		ret = 0;
+	}
+
+	return ret;
+}
+
+static inline void jumps_reloc(uint32_t loc, int32_t *offset)
+{
+	*offset >>= SHIFT_BRANCH_OFFSET;
+	*offset &= MASK_BRANCH_OFFSET;
 
-	addend <<= ARM_BL_BLX_ADDEND_SIZE;
-	addend |= upper & ARM_BL_BLX_ADDEND_MASK;
-	addend <<= ARM_BL_BLX_ADDEND_SIZE;
-	addend |= lower & ARM_BL_BLX_ADDEND_MASK;
+	*(uint32_t *)loc &= OPCODE2ARMMEM(MASK_BRANCH_COND|MASK_BRANCH_101|MASK_BRANCH_L);
+	*(uint32_t *)loc |= OPCODE2ARMMEM(*offset);
+}
+
+static int jumps_handler(elf_word reloc_type, uint32_t loc,
+				uint32_t sym_base_addr, const char *sym_name, int32_t *offset)
+{
+	int ret;
+
+	ret = jumps_decode(reloc_type, loc, sym_base_addr, sym_name, offset);
+	if (!ret) {
+		jumps_reloc(loc, offset);
+	}
+
+	return ret;
+}
+
+static void movs_handler(elf_word reloc_type, uint32_t loc,
+			uint32_t sym_base_addr, const char *sym_name, int32_t *offset)
+{
+	uint32_t tmp;
+
+	*offset = tmp = MEM2ARMOPCODE(*(uint32_t *)loc);
```

Comment:
Offset is not used outside of the handler and could be made local to this block avoiding pointer derefencing